boofuzz/socket_connection.py

```diff
@@ -165,13 +167,23 @@ def open(self):
 
         # if SSL is requested, then enable it.
         if self.proto == "ssl":
-            if self.server:
-                ssl_sock = ssl.wrap_socket(self._sock, keyfile=self.keyfile, certfile=self.certfile,
-                                                      server_side=True)
-                self._sock = ssl_sock
-            else:
-                ssl_sock = ssl.wrap_socket(self._sock)
-                self._sock = ssl_sock
+            if not self.sslcontext:
+                # If we are the SSL client and user did not give us a SSLContext,
+                # then we just use a default one. However, we only can verify the
+                # identity of the SSL server, if we have been given a server_hostname.
+                if not self.server and not self.sslcontext:
+                    self.sslcontext = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
+                    if self.server_hostname:
+                        self.sslcontext.check_hostname = True
+                        self.sslcontext.verify_mode = ssl.CERT_REQUIRED
+                    else:
+                        # Only option left would be to accept any certificates.
+                        # This would left us vulnerable to man-in-the-middle attacks.
+                        # We won't allow that and the user has to remove the assert below.
+                        assert(False)
+                        self.sslcontext.check_hostname = False
+                        self.sslcontext.verify_mode = ssl.CERT_NONE
+            self._sock = self.sslcontext.wrap_socket(self._sock, server_side=self.server, server_hostname=self.server_hostname)
```

Comment:
Let's keep `if self.server: ... else: ...` approach. I like that it makes very clear what the behavior of the client is and what the behavior of the server is.

For this PR, are you trying to address the server behavior or client behavior?

Reply:
This PR is for both: SSL server and client fuzzing. I do not need to care about the value of ```self.server```, because the ```wrap_socket``` method is exactly the same for both cases. See simplified code in next commit.